tiny_cnn/core/backend_avx.h

```diff
@@ -0,0 +1,285 @@
+    All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+    * Redistributions of source code must retain the above copyright
+    notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+    notice, this list of conditions and the following disclaimer in the
+    documentation and/or other materials provided with the distribution.
+    * Neither the name of the <organization> nor the
+    names of its contributors may be used to endorse or promote products
+    derived from this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+    EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY
+    DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+    (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+    ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#pragma once
+
+#include "tiny_cnn/core/backend.h"
+
+#include "tiny_cnn/core/kernels/tiny_conv2d_kernel.h"
+#include "tiny_cnn/core/kernels/tiny_conv2d_back_kernel.h"
+#include "tiny_cnn/core/kernels/tiny_deconv2d_kernel.h"
+#include "tiny_cnn/core/kernels/tiny_deconv2d_back_kernel.h"
+#include "tiny_cnn/core/kernels/tiny_maxpool_kernel.h"
+#include "tiny_cnn/core/kernels/tiny_fully_connected_kernel.h"
+
+#ifndef CNN_USE_AVX
+#error Advanced Vector Extensions required.
+#endif
+
+#include "tiny_cnn/core/kernels/avx_conv2d_kernel.h"
+#include "tiny_cnn/core/kernels/avx_conv2d_back_kernel.h"
+#include "tiny_cnn/core/kernels/avx_deconv2d_kernel.h"
+#include "tiny_cnn/core/kernels/avx_deconv2d_back_kernel.h"
+#include "tiny_cnn/core/kernels/avx_maxpool_kernel.h"
+#include "tiny_cnn/core/kernels/avx_fully_connected_kernel.h"
+
+namespace tiny_cnn {
+namespace core {
+
+class avx_backend : public backend {
+ public:
+    // context holds solution-dependent parameters
+    // context should be able to hold any types of structures (like boost::any)
+
+    // convolution
+    avx_backend(conv_params* params,
+                std::function<void(const vec_t&, int)> f1,
+                std::function<void(const vec_t&, vec_t&)> f2,
+                std::function<void(const vec_t&, const vec_t&, vec_t&)> f3,
+                std::vector<conv_layer_worker_specific_storage>* ptr)
+      : params_c_(params)
```

Comment:
you're right. However, just realized that backend_tiny is wrong.
Could you fix that? Basically as in dev README says, we follow 4 tabs convention